Bottom: 4d789e9547b4b96147504ed4747f7c4562d86088
Top:    2d2d02544f91fa852dbb6dd0a303180d603d93a3
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-10-01 10:49:31 +0100

Refresh of play-with-glsl-shader-for-rend

---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 96d733c..876f5a8 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -814,6 +814,12 @@ load_built_in_shaders (void)
   circular_program = hidgl_shader_new ("circular_rendering", NULL, circular_fs_source);
 }
 
+static void
+unload_built_in_shaders (void)
+{
+  hidgl_shader_free (circular_program);
+}
+
 void
 hidgl_init (void)
 {
@@ -843,6 +849,8 @@ hidgl_init (void)
     goto done;
   }
 
+  load_built_in_shaders ();
+
 done:
   done_once = true;
 }
@@ -856,7 +864,7 @@ hidgl_start_render (void)
   in_context = true;
   hidgl_init ();
   hidgl_init_triangle_array (&buffer);
-  load_built_in_shaders ();
+//  load_built_in_shaders ();
   hidgl_shader_activate (circular_program);
 }
 
@@ -868,6 +876,7 @@ hidgl_finish_render (void)
 
   hidgl_finish_triangle_array (&buffer);
   hidgl_shader_activate (NULL);
+//  unload_built_in_shaders ();
   in_context = false;
 }
 
diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index f40db60..87b2d0b 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -2231,6 +2231,7 @@ ghid_render_pixmap (int cx, int cy, double zoom, int width, int height, int dept
   GdkGLConfig *glconfig;
   GdkPixmap *pixmap;
   GdkGLPixmap *glpixmap;
+  GdkGLContext *drawarea_glcontext;
   GdkGLContext* glcontext;
   GdkGLDrawable* gldrawable;
   view_data save_view;
@@ -2244,6 +2245,11 @@ ghid_render_pixmap (int cx, int cy, double zoom, int width, int height, int dept
   /* Setup rendering context for drawing routines
    */
 
+  /* NB: We share with the main rendering context so we can use the
+   *     same pixel shader etc..
+   */
+  drawarea_glcontext = gtk_widget_get_gl_context (gport->drawing_area);
+
   glconfig = gdk_gl_config_new_by_mode (GDK_GL_MODE_RGB     |
                                         GDK_GL_MODE_STENCIL |
                                         GDK_GL_MODE_SINGLE);
@@ -2251,7 +2257,8 @@ ghid_render_pixmap (int cx, int cy, double zoom, int width, int height, int dept
   pixmap = gdk_pixmap_new (NULL, width, height, depth);
   glpixmap = gdk_pixmap_set_gl_capability (pixmap, glconfig, NULL);
   gldrawable = GDK_GL_DRAWABLE (glpixmap);
-  glcontext = gdk_gl_context_new (gldrawable, NULL, FALSE, GDK_GL_RGBA_TYPE);
+  glcontext = gdk_gl_context_new (gldrawable, drawarea_glcontext,
+                                  FALSE, GDK_GL_RGBA_TYPE);
 
   /* Setup zoom factor for drawing routines */
